resource/resourceadapters/opener_test.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/juju/charm/v8"
+	charmresource "github.com/juju/charm/v8/resource"
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/charmstore"
+	"github.com/juju/juju/resource"
+	"github.com/juju/juju/resource/resourceadapters"
+	"github.com/juju/juju/resource/resourceadapters/mocks"
+	"github.com/juju/juju/state"
+)
+
+type OpenerSuite struct {
+	app                 *mocks.MockApplication
+	unit                *mocks.MockUnit
+	resources           *mocks.MockResources
+	resourceGetter      *mocks.MockResourceGetter
+	resourceOpenerState *mocks.MockResourceOpenerState
+}
+
+var _ = gc.Suite(&OpenerSuite{})
+
+func (s *OpenerSuite) TestOpenResource(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	fp, _ := charmresource.ParseFingerprint("38b060a751ac96384cd9327eb1b1e36a21fdb71114be07434c0cc7bf63f6e1da274edebfe76f65fbd51ad2f14898b95b")
+	res := resource.Resource{
+		Resource: charmresource.Resource{
+			Meta: charmresource.Meta{
+				Name: "wal-e",
+				Type: 1,
+			},
+			Origin:      2,
+			Revision:    0,
+			Fingerprint: fp,
+			Size:        0,
+		},
+		ApplicationID: "postgreql",
+	}
+	s.expectCacheMethods(res)
+	s.resourceGetter.EXPECT().GetResource(gomock.Any()).Return(charmstore.ResourceData{
+		ReadCloser: nil,
+		Resource:   res.Resource,
+	}, nil)
+
+	opened, err := s.newOpener().OpenResource("wal-e")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(opened.Resource, gc.DeepEquals, res)
+}
+
+func (s *OpenerSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+	s.unit = mocks.NewMockUnit(ctrl)
+	s.app = mocks.NewMockApplication(ctrl)
+	s.resourceGetter = mocks.NewMockResourceGetter(ctrl)
+	s.resources = mocks.NewMockResources(ctrl)
+	s.resourceOpenerState = mocks.NewMockResourceOpenerState(ctrl)
+
+	s.unit.EXPECT().ApplicationName().Return("postgresql")
+	s.unit.EXPECT().Application().Return(s.app, nil)
+	curl, _ := charm.ParseURL("postgresql")
+	s.unit.EXPECT().CharmURL().Return(curl, false)
+	s.expectCharmOrigin()
+
+	return ctrl
+}
+
+func (s *OpenerSuite) expectCharmOrigin() {
+	rev := 0
+	s.app.EXPECT().CharmOrigin().Return(&state.CharmOrigin{
+		Source:   "charm-hub",
+		Type:     "charm",
+		Revision: &rev,
+		Channel:  &state.Channel{Risk: "stable"},
+		Platform: &state.Platform{
+			Architecture: "amd64",
+			OS:           "ubuntu",
+			Series:       "focal",
+		},
+	})
+}
+
+func (s *OpenerSuite) expectCacheMethods(res resource.Resource) {
+	s.resources.EXPECT().OpenResourceForUniter(gomock.Any(), gomock.Any()).Return(resource.Resource{}, ioutil.NopCloser(bytes.NewBuffer([]byte{})), errors.NotFoundf("wal-e"))
+	s.resources.EXPECT().GetResource("postgresql", "wal-e").Return(res, nil)
+	//s.resources.EXPECT().SetResource("postgresql", "postgresql/0", gomock.Any(), gomock.Any()).Return(res, nil)
+	s.resources.EXPECT().SetResource(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(res, nil)
```

Comment:
Something is wrong here.  The opener is somewhere calling the wrong version of SetResource, the one from EntityCache, which should be resourceCache.SetResource, instead of the one from Resources (into state).  

Gomock is allowing the issue because gomock.Any() can be none.  Anyone find the place?